Merge feature/code-review-refinements into develop

MarkoResic opened the pull request.

Commit:
Remove admin registration, change password encoder

- API no longer provides method to register as administrator.

Commit:
Add registration validation and error handling

Commit:
Add user creation validation, refactor error handling

- Stopped using custom exceptions in favor of generic exceptions with custom messages.

Commit:
Add method for getting all time votes

- Votes property of player details now refers to all-time votes.
- Voting with an invalid player ID now throws a NoSuchElementException.

Commit:
Implement refresh tokens

- Added API methods for logging out and refreshing access token.

MarkoResic committed:
Create README.md

oiramcro commented on src/main/kotlin/com/github/markoresic/footballvotingdemo/auth/AuthenticationService.kt:
```diff
@@ -1,32 +1,106 @@
+        val savedUser = try {
+            userService.createUser(user)
+        } catch (e: Exception) {
+            throw RuntimeException("User with e-mail already exists!")
```
This works now but It's very wired to trow RuntimeException here.  It's much safer not even try to save the user and check if email exists and return a custom exception to that. 

On top of that when you try to register a duplicate user you get generic 400 error with non-JSON body "User with e-mail already exists!" 

Please return json object as error and read this:

https://stackoverflow.com/questions/9269040/which-http-response-code-for-this-email-is-already-registered

and this:

https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-p2-semantics-18#section-7.4.10

oiramcro commented on src/main/kotlin/com/github/markoresic/footballvotingdemo/controller/VoteController.kt:
```diff
@@ -35,7 +37,7 @@ class VoteController(
             ResponseEntity.status(HttpStatus.CREATED).build()
         } else {
-            ResponseEntity.status(HttpStatus.BAD_REQUEST).build()
+            throw RuntimeException("You already voted today!")
```
This works now, but error body is not JSON

oiramcro commented on src/main/kotlin/com/github/markoresic/footballvotingdemo/controller/PlayerController.kt:
```diff
@@ -17,19 +18,36 @@ class PlayerController(
     }
 
     @GetMapping("/search/{searchTerm}")
     fun getPlayersBySearchTerm(@PathVariable searchTerm: String): List<PlayerListItemResponse> =
```
This is OK, it works. But usefully search terms are considered Query Params. 

Read more here: 

https://stackoverflow.com/questions/5020704/how-to-design-restful-search-filtering

oiramcro commented on src/main/kotlin/com/github/markoresic/footballvotingdemo/controller/VoteController.kt:
```diff
@@ -24,7 +26,7 @@ class VoteController(
 
         if (!playerService.playerExistsById(playerId)) {
-            return ResponseEntity.status(HttpStatus.BAD_REQUEST).build()
+            throw NoSuchElementException("Player not found!")
```
As in other examples error is not API friendly and does not return correct code and response body:

here is an example of ErrorHandler with @ControllerAdvice annotation which you can register for you Controler.    

Where error is a simple object 

data class SimpleApiError(val code: Int, val message: String, val type: String? = null)

oiramcro commented on src/main/kotlin/com/github/markoresic/footballvotingdemo/auth/RegisterRequest.kt:
```diff
@@ -1,10 +1,14 @@
+    @field:NotBlank(message = "Email must not be blank")
+    @field:Email(message = "Email must be valid")
     val email: String,
+    @field:NotBlank(message = "Password must not be blank")
```
What if you want to implement custom logic for password, let's say one special char and longer then 8? can you modify this?

oiramcro commented:
Please resolve Merge conflicts

Commit:
Refactor error handling, fix bugs

- Error handling methods now return JSON in response body.
- Replaced path variables with request parameters in relevant API methods.
- Added password validation for user registration.

MarkoResic committed:
Merge branch 'develop' into feature/code-review-refinements

Commit:
Revert "Merge branch 'develop' into feature/code-review-refinements"

- This reverts commit b06c1bb.

Commit:
Resolve conflict resolution with 02c3bf47

- Accepted proposed property name changes from "Fix: Fix some wired things in order to generate merge conflicts"
